File: problem98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
                                            
                                                15
                                /                           \
                            6                                   20
                        /       \                           /       \
                    3               9                   12             25
                /       \          /    \
                1       8        7       10
                
                        [-inf, 15]
                    [-inf, 6]   [6, 15]
                [-inf,3] [3, 6] []
            [-inf,1]
                                            
"""

class Solution:
    def isValidBST(self, node: Optional[TreeNode], low=float('-inf'), high=float('inf')) -> bool:
        
        if not low<node.val<high:
            return False
            
        return (not node.left or self.isValidBST(node.left, low, node.val)) and (not node.right or self.isValidBST(node.right, node.val, high))

File: problem98.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public bool IsValidBST(TreeNode root) {
        return CheckBST(Int64.MinValue, Int64.MaxValue, root);
    }
    
    bool CheckBST(long minValue, long maxValue, TreeNode node)
    {
        if(node==null)
            return true;
        
        // Eventhough current root may not be bounded to [minvalue, maxvalue] and return true, the recursive calls to left and right children will return false if the values are not bounded to [minvalue, maxvalue]
        
        return CheckBST(minValue, node.val, node.left)
            && CheckBST(node.val, maxValue, node.right)
            && node.val<maxValue
            && node.val>minValue;
    }
}


//Time:O(N)
//Space: O(N)
